Integrate the work done by @jdeneef back to this repo

tomer-w opened the pull request:
This will allow Raspberry Pi 5 support and a few more advanced tweaks.
Based on the discussion here: https://github.com/thecode/ha-rpi_gpio/issues/278

tomer-w commented:
@jdeneef, @raboof, FYI.

tomer-w commented:
Closed by mistake.

thecode commented:
@tomer-w when we discussed about making a PR to update this repo my intention was not just to merge another repo into this one, otherwise I could do it without a PR.

Please remove anything which is not necessary from this PR, docs, license, json file updates should not be in this PR, focus on only code changes if other changes are needed please move them to a different PR.
In addition if you want this PR to reviewed faster try to split it into multiple PRs, I can create a branch so we merge PRs into that branch and merge it to main when ready, otherwise the review process will take time.

tomer-w commented:
Sure, feel free to create a branch. It will make it easier.
This PR is not only merging different repo, it is aligning it back with the original yaml format so it is backwards compatible.
Most of the code is coming from the source repo so there is enough data that it is actually working. I think most of the focus should be on the docs, json, etcs. 
Anyhow I published it that we will have something to work with, branch will be the best solution.

thecode commented:
> Sure, feel free to create a branch. It will make it easier. 

https://github.com/thecode/ha-rpi_gpio/tree/gpiod-rpi5

tomer-w commented:
PR opened. Can you give me write access to that branch?

tomer-w commented:
I moved this PR to the new branch. Please approve and give me write access so we can iterate over it.

thecode commented:
I think my comment was misunderstood. I asked to split this PR into several PRs made to the new branch, not that you merge content into this branch without review.

tomer-w commented:
I removed from the PR all the auxiliary changes. We now have only the actual code, the documentation which states the new parameters and a couple more super minor changes. It is only 11 files, 7 of them are code files. Hope this is small enough we can review.

thecode commented:
Please split the docs into a different PR, I will start reviewing code changes

thecode commented on custom_components/rpi_gpio/const.py:
```diff
@@ -0,0 +1,4 @@
+
+DOMAIN = "gpiod"
```
```suggestion
DOMAIN = "rpi_gpiod"
```
I know this is now supporting other platforms, but the name needs to match the intergration

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
+            return
+        self._hub.turn_on(self._relay_port)
+        self.is_closing = True
+        # self.is_closed = None
```
```suggestion
```
Please remove all dead code

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/hub.py:
```diff
@@ -0,0 +1,206 @@
+        # setup lines
+        self.update_lines()
+
+        # read initial states for sensors
+        # for port in self._config:
+            # self._entities[port].update()
+
```
Dead code

tomer-w replied:
fixed

thecode commented on custom_components/rpi_gpio/manifest.json:
```diff
@@ -1,10 +1,11 @@
   "domain": "rpi_gpio",
   "name": "Raspberry Pi GPIO",
-  "codeowners": ["@thecode"],
+  "codeowners": [ "@thecode", "@jdeneef", "@tomer-w" ],
```
This will be done in a separate PR

tomer-w replied:
ok

thecode commented on custom_components/rpi_gpio/switch.py:
```diff
@@ -1,160 +1,116 @@
+    def update(self):
+        self.is_on = self._hub.update(self._port)
+        self.schedule_update_ha_state(False)
+        
```
```suggestion
```

tomer-w replied:
fixed

tomer-w replied:
fixed.

thecode reviewed (changes requested):
- Remove docs from this PR
- Remove all commented out code
- make sure domain name doesn't change-
-  When hub is offline we should handle this better, not by just logging an error

tomer-w commented:
I fixed all the CR comments, re-tested the integration and removed the minor non code changes other than the documentation itself as I dont want to lose the changes. Can we finish the CR as is? I can spend time about splitting the single MD file but it is easier to check it in into the branch and then iterate on it.

thecode commented on README.md:
````diff
@@ -183,4 +208,12 @@ switch:
       - port: 17
         name: "Speaker Relay"
-```
+
+
+# Add Debug info and issue reporting
+*Before* reporting issues please add this to your `configuration.yaml` `logger` section, check logs and report issue adding logging.
+```yaml
+logger:
+  default: info
+  logs:
+    custom_components.rpi_gpio: debug
````
Please link to https://www.home-assistant.io/docs/configuration/troubleshooting/#enabling-debug-logging instead. There is no need to use YAML anymore for debug and the UI functionality automatically download the log upon disable.

tomer-w replied:
done

thecode commented on hacs.json:
```diff
@@ -1,4 +1,5 @@
 {
   "name": "Raspberry Pi GPIO",
-  "homeassistant": "2022.7.0"
+  "homeassistant": "2022.7.0",
+  "render_readme": true
```
```suggestion
  "homeassistant": "2022.7.0"
```
`render_readme` is no longer supported by `HACS`

tomer-w replied:
done, but dont you want to bump up the version?

thecode replied:
We will bump it before the release

thecode reviewed (changes requested):
Reviewed only the docs for now

tomer-w commented:
Fixed all current doc comments. My only question was regarding the versioning which I assumed we should bump.

jdeneef commented:
@tomer-w quick check on your changes shows that you are removing all config validations and parameters names reverting it back to rpi-gpio terminology? I'm a bit puzzled why you would do that? I made those changes on purpose to get closer to the actual used libgpiod library for anybody who really needs to understand what the parameter actually does, not having to reverse engineer the code to find out what parameters is actually set?

tomer-w commented:
I fully understand your motives and it makes sense from that perspective but as the rpi-gpio integration is heavily used, I didn't want to break things for current users or stress them with documentation which says that those old configuration can be deprecated.
At least this was my thinking, but I'm open to any feedback.

jdeneef commented:
ok, clear, thanx for explaining! I already had the plan to simply keep both since I did find a nice way without code duplication in the config validation, just hadn't updated docs just yet. So you could mention both new and old config options in one row, keep the config validation as is and remove all 'backwards compatibility' stuff, that could be a fine option in the middle? BTW I did not know it was still used? It stopped working for me like 8 months ago when my distro moved to 6..1.69. So fair enough to make transfer as simple as possible.

tomer-w commented:
Sound good. I'm waiting for @thecode to merge it into the branch. Than we can work on it together to make sure it is backwards compatible and still can be used with the native naming you used.

tomer-w commented:
And yes, the integration still works as far as you didn't move to RP5.

thecode commented on custom_components/rpi_gpio/binary_sensor.py:
```diff
@@ -1,141 +1,89 @@
+    def __init__(self, hub, name, port, unique_id, active_low, bias, debounce):
+        _LOGGER.debug(f"GPIODBinarySensor init: {port} - {name} - {unique_id}")
+        self._hub = hub
+        self.name = name
```
```suggestion
        self._attr_name = name
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
+                 state_port, state_bias, state_active_low, unique_id):
+        _LOGGER.debug(f"GPIODCover init: {relay_port}:{state_port} - {name} - {unique_id} - {relay_time}")
+        self._hub = hub
+        self.name = name
```
```suggestion
        self._attr_name = name
```

tomer-w replied:
fixed.

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/switch.py:
```diff
@@ -1,160 +1,114 @@
+    def __init__(self, hub, name, port, unique_id, active_low, bias, drive, persistent):
+        _LOGGER.debug(f"GPIODSwitch init: {port} - {name} - {unique_id} - active_low: {active_low} - bias: {bias} - drive: {drive}")
+        self._hub = hub
+        self.name = name
```
```suggestion
        self._attr_name = name
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
-        """Return true if cover is closed."""
-        return self._state != self._invert_state
+        self.is_closed = self._hub.update(self._state_port)
+        self.schedule_update_ha_state(False)
```
```suggestion
```
This should cause update to happen twice, please verify if it is really needed

thecode commented on custom_components/rpi_gpio/binary_sensor.py:
```diff
@@ -1,141 +1,89 @@
-        """Update the GPIO state."""
-        self._state = read_input(self._port)
+        self.is_on = self._hub.update(self._port)
+        self.schedule_update_ha_state(False)
```
```suggestion
```
This should cause update to happen twice, please verify if it is really needed

tomer-w replied:
I didn't notice something happens twice. @thecode, why do you suspect that? Maybe I'm missing something.

thecode replied:
You can't notice if something update twice, only if you enable debug logging for core (which is really hard to read later) or check the DB.
What you can do is remove this line, if everything is updated correctly (for example when you change the input state it updates in HA) it means it is not necessary, from looking at Home Assistant core code this should not be needed, please verify.

thecode commented on custom_components/rpi_gpio/switch.py:
```diff
@@ -1,160 +1,114 @@
\ No newline at end of file
+    def update(self):
+        self.is_on = self._hub.update(self._port)
+        self.schedule_update_ha_state(False)
```
```suggestion
```
This should cause update to happen twice, please verify if it is really needed

tomer-w replied:
I didn't notice something happens twice. @thecode, why do you suspect that? Maybe I'm missing something.

thecode commented on custom_components/rpi_gpio/binary_sensor.py:
```diff
@@ -1,141 +1,89 @@
-        self._state = await self.hass.async_add_executor_job(read_input, self._port)
-        self.async_write_ha_state()
+class GPIODBinarySensor(BinarySensorEntity):
+    should_poll = False
```
```suggestion
    _attr_should_poll = False
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
+    async_add_entities(covers)
+
+class GPIODCover(CoverEntity):
+    should_poll = False
```
```suggestion
    _attr_should_poll = False
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/hub.py:
```diff
@@ -0,0 +1,192 @@
+
+class Hub:
+
+    manufacturer = "ha_gpiod"
```
```suggestion
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/switch.py:
```diff
@@ -1,160 +1,114 @@
+
+
+class GPIODSwitch(SwitchEntity, RestoreEntity):
+    should_poll = False
```
```suggestion
    _attr_should_poll = False
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/switch.py:
```diff
@@ -1,160 +1,114 @@
+    def __init__(self, hub, name, port, unique_id, active_low, bias, drive, persistent):
+        _LOGGER.debug(f"GPIODSwitch init: {port} - {name} - {unique_id} - active_low: {active_low} - bias: {bias} - drive: {drive}")
+        self._hub = hub
+        self.name = name
```
```suggestion
        self._attr_name = name
```

thecode commented on custom_components/rpi_gpio/binary_sensor.py:
```diff
@@ -1,141 +1,89 @@
+        _LOGGER.debug(f"GPIODBinarySensor init: {port} - {name} - {unique_id}")
+        self._hub = hub
+        self.name = name
+        self.unique_id = unique_id
```
```suggestion
        self._attr_unique_id = unique_id
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
+        self._state_port = state_port
+        self._state_bias = state_bias
+        self._start_active_low = state_active_low
+        self.unique_id = unique_id
```
```suggestion
        self._attr_unique_id = unique_id
```

tomer-w replied:
fixed.

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
-    def is_closed(self):
-        """Return true if cover is closed."""
-        return self._state != self._invert_state
+        self.is_closed = self._hub.update(self._state_port)
```
```suggestion
        self._attr_is_closed = self._hub.update(self._state_port)
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
+            # closing stopped
+            return
+        self._hub.turn_off(self._relay_port)
+        self.is_closing = False
```
```suggestion
        self._attr_is_closing = False
```

tomer-w replied:
fixed.

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
+        if self.is_closed:
+            return
+        self._hub.turn_on(self._relay_port)
+        self.is_closing = True
```
```suggestion
        self._attr_is_closing = True
```

tomer-w replied:
fixed.

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
-            self._trigger()
+            return
+        self._hub.turn_on(self._relay_port)
+        self.is_opening = True
```
```suggestion
        self._attr_is_opening = True
```

tomer-w replied:
fixed.

thecode replied:
Where is it fixed?

tomer-w replied:
Will submit in 2 hours.

thecode reviewed (changes requested):
Please go over all of the attributes and make sure they are set via the `_attr_` and not directly.

tomer-w commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,100 @@
+            return
+        self._hub.turn_on(self._relay_port)
+        self.is_opening = True
+        # self.is_closed = None
```
fixed.

tomer-w commented:
Merged fixes for most of the comments.

tomer-w commented on custom_components/rpi_gpio/hub.py:
```diff
@@ -0,0 +1,192 @@
+
+class Hub:
+
+    manufacturer = "ha_gpiod"
```
fixed.

tomer-w commented:
I think I handled all the CR comments other than the question about update happens twice which I didn't understand. Let me know if anything else is missing.

thecode commented on custom_components/rpi_gpio/hub.py:
```diff
@@ -0,0 +1,189 @@
+} 
+
+class Hub:
+
```
```suggestion
```

thecode commented on custom_components/rpi_gpio/hub.py:
```diff
@@ -0,0 +1,188 @@
+
+        if not self._online:
+            _LOGGER.error("No gpio device detected, bailing out")
+            raise IntegrationError("No gpio device detected")
```
```suggestion
            raise HomeAssistantError("No gpio device detected")
```
`IntegrationError` is internal for HA, integrations should raise `HomeAssistantError`

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/hub.py:
```diff
@@ -0,0 +1,188 @@
+
+from homeassistant.core import HomeAssistant
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP, EVENT_HOMEASSISTANT_START
+from homeassistant.exceptions import IntegrationError
```
```suggestion
from homeassistant.exceptions import HomeAssistantError
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,99 @@
-        """Close the cover."""
-        if not self.is_closed:
-            self._trigger()
+        if self._attr_is_closed:
```
```suggestion
        if self.is_closed:
```
I know this is confusing, but writing should use the `_attr_` and read direct from the property. This allow HA to cache values.

tomer-w replied:
You always learn something new. Where do you learn that from?

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,99 @@
+        self._attr_is_closing = True
+        self.schedule_update_ha_state(False)
+        sleep(self._relay_time)
+        if not self._attr_is_closing:
```
```suggestion
        if not self.is_closing:
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,99 @@
-        """Open the cover."""
-        if self.is_closed:
-            self._trigger()
+        if not self._attr_is_closed:
```
```suggestion
        if not self.is_closed:
```

tomer-w replied:
fixed.

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,99 @@
+        self._attr_is_opening = True
+        self.schedule_update_ha_state(False)
+        sleep(self._relay_time)
+        if not self._attr_is_opening:
```
```suggestion
        if not self.is_opening:
```

thecode commented on custom_components/rpi_gpio/cover.py:
```diff
@@ -50,88 +53,99 @@
+        self.update()
+
+    def stop_cover(self, **kwargs):
+        if not (self._attr_is_closing or self._attr_is_opening):
```
```suggestion
        if not (self.is_closing or self.is_opening):
```

tomer-w replied:
fixed.

thecode reviewed (changes requested):
Minor last comments, please fix and we should be ready to merge this

tomer-w commented:
Latest CR comments applied.

tomer-w commented:
@thecode, going back to the update twice issue, your assumption is that:
```self.is_on = self._hub.update(self._port)```
will also get HA to update this state so:
```self.schedule_update_ha_state(False)```
is not needed?

tomer-w commented:
@jdeneef , tagging you for the issue about the above issue.

thecode commented:
> I can confirm from testing that the `update` function in `switch` will not get called at all as flipping the switch does not trigger fd update. It is called only for `binary_sensor` updates when the `libgpio` sense change in the voltage.

My confusing was becasue `update` is also a method used by HA to call when it needs to update an entity and in this case there is no need to call `self.schedule_update_ha_state`, however if it is called from another source you do need to call `self.schedule_update_ha_state`.

tomer-w commented:
I had the same confusion initially and I think we should rename our ```update``` to something which will make it clear it is related only to our integration.

tomer-w commented:
@jdeneef, I think your checking from last week broke cover: https://github.com/jdeneef/ha_gpiod/commit/ef70c3379822f280c81ed0ad254bb9af15288017
In the cover case entity is null and this line breaks:
```
entity.is_on = True if value ^ active_low else False
```
@thecode, we can't integrate it till this is fixed.

thecode commented:
I have converted the issue to a discussion, we can continue the planning on the discussion: https://github.com/thecode/ha-rpi_gpio/discussions/285
